src/init.ml

```diff
@@ -0,0 +1,198 @@
+open! Stdune
+open! Import
+
+module Project = struct
+  module Kind = struct
+    type t =
+      | Library
+      | Executable
+
+    let to_string = function
+      | Library -> "library"
+      | Executable -> "executable"
+  end
+end
+
+module File = struct
+  type t =
+    { path: Path.t
+    ; name: string
+    ; content: string }
+
+  let make path name content =
+    {path; name; content}
+
+  let full_path {path; name; _} =
+    Path.(append path (of_string name))
+
+  let create file =
+    let path = full_path file in
+    Io.write_file ~binary:false path file.content
+end
+
+module type InitContext = sig
+  val kind : Project.Kind.t
+  val pname : string
+  val dir : Path.t
+end
+
+module type ProjectDir = sig
+  val dir : Path.t
+  val files : File.t list
+end
+
+module Proj (C : InitContext) : ProjectDir = struct
+  let dir = C.dir
+  let make_file = File.make C.dir
+
+  let dune_project =
+    let lang_stanza = Dune_project.lang_stanza () in
+    let name = Dune_project.filename in
+    let content = Printf.sprintf "\
+%s
+(name %s)
+(version 0.1)
+" lang_stanza C.pname
+    in
+    make_file name content
+
+  let package_opam =
+    let name = Printf.sprintf "%s.opam" C.pname in
+    make_file name ""
+
+  let files = [dune_project; package_opam]
+end
+
+module Lib (C : InitContext) : ProjectDir = struct
+  let dir = Path.(append C.dir (of_string "src"))
+  let make_file = File.make dir
+
+  let lib_dune =
+    let name = "dune" in
+    let content = Printf.sprintf "\
+;; Uncomment stanzas to include their functionality
+(library
+  (name %s)
+  (public_name %s)
+  ;; (inline_tests)
+  ;; (libraries core)
+  ;; (preprocess (pps ppx_inline_test))
+)
+" C.pname C.pname
+    in
+    make_file name content
+
+  let lib_ml =
+    let name = "lib.ml" in
+    let content = {|let hello name = Printf.sprintf "Hello, %s" name
+|} in
+    make_file name content
+
+  let files = [lib_dune; lib_ml]
+end
+
+module Bin (C : InitContext) : ProjectDir = struct
+  let dir = Path.(append C.dir (of_string "bin"))
+  let make_file = File.make dir
+
+  let bin_dune =
+    let name = "dune" in
+    let content = Printf.sprintf "\
+(executable
+  (public_name %s)
+  (name main)
+  (libraries %s))
+" C.pname C.pname
+    in
+    make_file name content
+
+  let bin_ml =
+    (* The depreciated [String.capitalize] is required over the preferred
+       [String.capitalize_ascii] to support OCaml 4.02.3 *)
+    let lib_module_name = String.capitalize C.pname in
+    let name = "main.ml" in
+    let content =
+      Printf.sprintf
+        {|let () = print_endline (%s.Lib.hello "project %s")
+|}
+        lib_module_name C.pname
+    in
+    make_file name content
+
+  let files = [bin_dune; bin_ml]
+end
+
+let make_project_dir_exn (module D : ProjectDir) =
+  if Path.exists D.dir then
```

Comment:
Isn't it simpler to have `mkdir -p` behavior here? If it the directory is there - great, let's proceed.